File: leetcode/cpp/90. 子集 II.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> ans;
vector<int> t;

void dfs(int startIndex, vector<int> nums) {
    if (t.size() == nums.size()) {
        return;
    }
    for (int i = startIndex; i < nums.size(); i++) {
        if (i > startIndex && nums[i] == nums[i-1]) {
            continue;
        }
        t.push_back(nums[i]);
        ans.push_back(t);
        dfs(i+1, nums);
        t.pop_back();
    }
}

vector<vector<int>> subsetsWithDup(vector<int>& nums) {
    dfs(0, nums);
    return ans;
}

int main() {
    vector<int> nums({1, 2, 2});
    vector<vector<int>> ret = subsetsWithDup(nums);
    
}

File: leetcode/golang/problems/90. 子集 II.go
package problems

import "sort"

func SubsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	ans := [][]int{}
	ans = append(ans, []int{})
	array := []int{}
	var dfs func(int)
	dfs = func(startIndex int) {
		if len(array) == len(nums) {
			return
		}
		for i := startIndex; i < len(nums); i++ {
			if i > startIndex && nums[i] == nums[i-1] {
				continue
			}
			array = append(array, nums[i])
			temp := make([]int, len(array))
			copy(temp, array)
			ans = append(ans, temp)
			dfs(i+1)
			array = array[:len(array)-1]
		}
	}
	dfs(0)

	return ans
}